src/Mod/Sketcher/Gui/SnapManager.cpp

```diff
@@ -0,0 +1,309 @@
+ *                                                                         *
+ *   This library  is distributed in the hope that it will be useful,      *
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
+ *   GNU Library General Public License for more details.                  *
+ *                                                                         *
+ *   You should have received a copy of the GNU Library General Public     *
+ *   License along with this library; see the file COPYING.LIB. If not,    *
+ *   write to the Free Software Foundation, Inc., 59 Temple Place,         *
+ *   Suite 330, Boston, MA  02111-1307, USA                                *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "PreCompiled.h"
+#ifndef _PreComp_
+#include <QApplication>
+#endif  // #ifndef _PreComp_
+
+#include <Mod/Sketcher/App/SketchObject.h>
+
+#include "SnapManager.h"
+#include "ViewProviderSketch.h"
+
+
+using namespace SketcherGui;
+using namespace Sketcher;
+
+/************************************ Attorney *******************************************/
+
+inline int ViewProviderSketchSnapAttorney::getPreselectPoint(const ViewProviderSketch& vp)
+{
+    return vp.getPreselectPoint();
+}
+
+inline int ViewProviderSketchSnapAttorney::getPreselectCross(const ViewProviderSketch& vp)
+{
+    return vp.getPreselectCross();
+}
+
+inline int ViewProviderSketchSnapAttorney::getPreselectCurve(const ViewProviderSketch& vp)
+{
+    return vp.getPreselectCurve();
+}
+
+/**************************** ParameterObserver nested class *****************************/
+SnapManager::ParameterObserver::ParameterObserver(SnapManager& client) : Client(client)
+{
+    initParameters();
+    subscribeToParameters();
+}
+
+SnapManager::ParameterObserver::~ParameterObserver()
+{
+    unsubscribeToParameters();
+}
+
+void SnapManager::ParameterObserver::initParameters()
+{
+    // static map to avoid substantial if/else branching
+    //
+    // key->first               => String of parameter,
+    // key->second              => Update function to be called for the parameter,
+    str2updatefunction = {
+        {"SnapToObjects",
+            [this](const std::string& param) {updateSnapToObjectParameter(param); }},
+        {"SnapAngle",
+            [this](const std::string& param) {updateSnapAngleParameters(param); }},
+    };
+
+    for (auto& val : str2updatefunction) {
+        auto string = val.first;
+        auto function = val.second;
+
+        function(string);
+    }
+}
+
+void SnapManager::ParameterObserver::updateSnapToObjectParameter(const std::string& parametername)
+{
+    (void)parametername;
+
+    ParameterGrp::handle hGrp = App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/Mod/Sketcher/General");

```

Comment:
hGrp shall be set as a const class static member variable, or return of a const static member function, so it is only defined once and then easy to reuse.

Reply:
const class static member variables have quite some lifetime:
https://stackoverflow.com/questions/1079623/what-is-the-lifetime-of-class-static-variables-in-c

This means that memory that is once allocated (on first use of the module) won't be recovered (for other purposes) until the program ends (like a life sentence). Sometimes, it can be better justified than others.

For parameters, when using an observer, there is no polling at all. Generally this points to non-intensive use. For the interrupt-like push of non-frequent changes, IMO a static does not fare well in this trade-off.

Reply:
Fair enough. Let's go for a member function then. :)

Reply:
What do you have in mind exactly?

Reply:
Something like
```
static ParameterGrp::handle SnapManager::ParameterObserver::getParameterGrpHandle()
{
    return App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/Mod/Sketcher/General");
}
```

Reply:
@abdullahtahiriyo can you please confirm if I implement the change offered by OpenBrain?

Reply:
The same function marked as static or non-static has practically equivalent performance. The only difference is that a member function takes an additional parameter internally, the pointer to the object on which it is called. Being this pointer unused, a compiler will optimize it.

So here we are talking about using a function to encapsulate the detail of the parameter. So, if we were to change the parameter string, we only have to change it once. That is the only advantage I could find to make a function.

So basically, there are three choices:
(a) normal inline private function (non static)  to encapsulate the detail. Even if not inline, the compiler will optimize this creation in place.
(b) a static non-inline private function.
(c) a free implementation only function.

The general choice of Scott Meyers is (c), see for example:
https://stackoverflow.com/questions/5989734/effective-c-item-23-prefer-non-member-non-friend-functions-to-member-functions

The logic behind is that you should not depend on what you do not need. You do not need access to members, so you do not need (a). You do not need access to static members, so you do not need (b).

Humbly (because I am a quite fervent Scott Meyers admirer), I would not do (c) here, as we do not need a public function and it disassociates the function from the class, and here that association is important. However, I thought important to mention it.

We could happily go along with (a) or (b) in this case, (b) may be better for the reasons explained above.

@PaddleStroke 
It makes sense. I am trying to explain the reasons of why we do some things or we don't. Please, try to understand why.

Also I do not hold the absolute truth. If any of you thinks I am mistaken, please come back with arguments.

Reply:
I will need to sit and think about that one because it seems I'm missing some understanding basis.
For now do I apply b) or leave as it currently is? I'm thinking that it's as I implemented in coinmanager. So is it good to have separate implementation rather than homogeneity?
Maybe it's best to change them all in another PR?

Reply:
As I said, it does have an advantage to use a function, as the parameter is not repeated and if need be changed, it has to be changed only in one place.

I would do two things: I will do the change (b) and then I will sit a think about what is written. It is not written to show off. It is written so that you learn and become better than me.

Reply:
Yes I understand, I'll do just that.